docs/spec/tuples.rst

```diff
@@ -0,0 +1,155 @@
+types such as ``tuple[int, *tuple[str, ...], str]`` -- a tuple type where the
+first element is guaranteed to be of type ``int``, the last element is
+guaranteed to be of type ``str``, and the elements in the middle are zero or
+more elements of type ``str``. The type ``tuple[*tuple[int, ...]]`` is
+equivalent to ``tuple[int, ...]``.
+
+If an unpacked ``*tuple[Any, ...]`` is embedded within another tuple, that
+portion of the tuple is bidirectionally type compatible with any tuple of
+any length.
+
+Only one unbounded tuple can be used within another tuple::
+
+    t1: tuple[*tuple[str], *tuple[str]]  # OK
+    t2: tuple[*tuple[str, *tuple[str, ...]]]  # OK
+    t3: tuple[*tuple[str, ...], *tuple[int, ...]]  # Type error
+    t4: tuple[*tuple[str, *tuple[str, ...]], *tuple[int, ...]]  # Type error
+
+An unpacked TypeVarTuple counts as an unbounded tuple in the context of this rule::
+
+    def func[*Ts](t: tuple[*Ts]):
+        t5: tuple[*tuple[str], *Ts]  # OK
+        t6: tuple[*tuple[str, ...], *Ts]  # Type error
+
+The ``*`` syntax requires Python 3.11 or newer. For older versions of Python,
+the ``typing.Unpack`` special form can be used:
+``tuple[int, Unpack[tuple[str, ...]], int]``.
+
+Unpacked tuples can also be used for ``*args`` parameters in a function
+signature: ``def f(*args: *tuple[int, str]): ...``. Unpacked tuples
+can also be used for specializing generic classes or type variables that are
+parameterized using a ``TypeVarTuple``. For more details, see
+:ref:`args_as_typevartuple`.
+
+
+Type Compatibility Rules
+------------------------
+
+Because tuple contents are immutable, the element types of a tuple are covariant.
+For example, ``tuple[int, int]`` is a subtype of ``tuple[float, complex]``.
+
+As discussed above, a homogeneous tuple of arbitrary length is equivalent
+to a union of tuples of different lengths. That means ``tuple[()]``, 
+``tuple[int]`` and ``tuple[int, *tuple[int, ...]]`` are all subtypes of
+``tuple[int, ...]``. The converse is not true; ``tuple[int, ...]``` is not a
+subtype of ``tuple[int]``.
+
+The type ``tuple[Any, ...]`` is bidirectionally compatible with any tuple::
+
+    def func(t1: tuple[int], t2: tuple[int, ...], t3: tuple[Any, ...]):
+        v1: tuple[int, ...] = t1  # OK
+        v2: tuple[Any, ...] = t1  # OK
+
+        v3: tuple[int] = t2  # Type error
+        v4: tuple[Any, ...] = t2  # OK
+
+        v5: tuple[float, float] = t3  # OK
+        v6: tuple[int, *tuple[str, ...]] = t3  # OK
+
+
+The length of a tuple at runtime is immutable, so it is safe for type checkers
+to use length checks to narrow the type of a tuple::
+
+    def func(val: tuple[int] | tuple[str, str] | tuple[int, *tuple[str], int]):
+        if len(val) == 1:
+            # Type can be narrowed to tuple[int].
+            reveal_type(val)  # tuple[int]
+        
+        if len(val) == 2:
+            # Type can be narrowed to tuple[str, str] | tuple[int, int].
```

Comment:
This seems incorrect, how do you get `tuple[int, int]`? Did you mean `tuple[int, *tuple[str, ...], int]` as the last element in the union?

Reply:
I think this is right. The `*tuple[str, ...]` expands to 0 or more elements. If the tuple is of length 2, that means the variadic part must be empty (because there must be an int at the start and one at the end).

Reply:
Yes that's what I'm saying it should be. The type of the argument is currently `tuple[int, *tuple[str], int]`, not `tuple[int, *tuple[str, ...], int]`

Reply:
Oh you're right, good catch.

Reply:
Yes, good catch. I caught the problem in the test suite, but I forgot to update the bug in the spec. Consider it fixed.